apis/condition_set.go

```diff
@@ -52,6 +52,17 @@ type ConditionManager interface {
 	// set to true.
 	IsHappy() bool
 
+	// IsUnknown looks at the happy condition and returns true if that condition is
```

Comment:
There is not a reason for the manager to use these methods.

Reply:
I checked the comment of ConditionManager.
It said "ConditionManager allows a resource to operate on its Conditions using higher order operations".  I think the methods I add falls into this category. I think it makes sense to add `IsUnknown()` and `isFalse()` into `ConditionManager` if we want to differentiate between status `Unknown` and 'False` for resources.

Reply:
In this https://github.com/knative/eventing/issues/2187, we talked about the standard of  True/False/Unknown status. This standard will be used in both serving and eventing. For this standard, we will need to differentiate between status `Unknown` and `False` for resources.

Reply:
@n3wscott  Besides `type conditionsImpl`,  do we have other integrators who implements the ConditionManager?

Reply:
> Could you share me with the example that #965 will break the integrations for all other knative resources?

ah ok you are correct, it is just the internal impl that does the impl of this interface.

Reply:
> For this standard, we will need to differentiate between status Unknown and False for resources.

We can already do this today. I think the helper methods make it a little unclear what is happening in the condition set. The Resource impl creates methods to help deal with the complexity, but I don't think it is the place for the condition manager to do this because what nil means and what true, false means for resources could differ. 

The bar for pkg code should be high and universally useful.

Reply:
Is there a reason why we couldn't introduce a new method called:
GetStatus() corev1.ConditionStatus

and in the Impl it just returns
r.GetCondition(r.happy).Status

Then, I _think_ (haven't thought this through) in our XXX_types.go (which means our Status duck) we should expose GetResourceStatus() or GetStatus() (which is wonky, so just spitballing here about the name) that would be more useful than the current IsReady() we have in our Status fields

Some random comments here:
Just to make sure I'm on the same page, when you say "We can already do this today.", do you mean that we can do this because we know if we've created either a Living or Batch condition set? Because they define which kind of a Condition you have (Ready, or Succeeded). If that's the reasoning (and I might have missed something), then I think having an explicit GetHappyCondition (Though, I'd probably rename that to be something like GetTopLevelCondition, or GetRootCondition, or something like that) would be much more clear. And, I think Living/Batch in the name does not make things much clearer either, they might as well be GetReadyConditionSet and GetSucceededConditionSet because those are the ConditionTypes they actually bake in (which we have to explain in the comments).
But, in any case, the ConditionSetManager is indeed meant only to make propagating things to the "happy" state easier, which it does but there are some assumptions built in that I personally think are little wonky. The second part of where this is used is the actual XXX_types where the resource Status fields expose the IsReady() by calling the IsHappy method in the ConditionSet, which does not give enough information to anything that's relying on propagating the status to the resource (because we need a tri-state, not boolean: True, False and Unknown).
For example, Sequence uses Subscriptions, and today it has to know which is the Happy condition in order to fetch the "happy" condition instead of asking for the top level. And while you can define something like a SubscriptionConditionReady = apis.ConditionReady
and pass that to the NewLivingConditionSet, it's only because the string values match (and we remove dupes) that they are the same thing (and if by mistake you create a NewBatchConditionSet and pass that to it, who knows what happens).
So, I think getting rid of the Happy in the name is a good thing and should probably be called GetTopLevelCondition or something like that for both ConditionManager as well as in our Status using that method to expose the actual ConditionStatus (or entire Top level condition) instead of boolean for various types, because I think you typically will just want to propagate (or check for ready) and having Is*** makes it harder.

Reply:
Yep. I agree that "GetTopLevelCondition" is more clear. Since we have this `"GetTopLevelCondition()` method, should we deprecate `IsHappy() bool` in ConditionManager?